### ISSUE TITLE
Html: Unhandled exception in ieframe.dll

### ISSUE THREAD

**Issue by dlaundon** [mity/mctrl#43]
Hi there, I use your library for the HTML control which is very useful, thanks!

We occasionally get reports of an unexplained exception in ieframe.dll that I could never replicate on my PC until now, and I can now get it to happen in a slightly modified example-html.exe.

Version: 0.11.1 (debug)
OS: Windows 10 version 1511 (OS Build 10586.633)
App: example-html.exe (modified)

I have modified the example-html project from the 0.11.1 source package (changed files attached):
* changed doc.html resource to contain embedded youtube video
* changed the back or forward toolbar buttons to close then reopen the HTML child control
* some minor changes to the main window to cope with missing HTML child control

Steps:
* Compile and run the modified example-html.exe (I've been using VS2015)
* Randomly click the back or forward button to destroy and recreate the HTML control (sometimes quickly, sometimes slowly)
* Eventually the crash will occur when:
  * WS_APP + 2 windows message is posted to an Internet Explorer_Hidden window
  * Some sort of delayed clean up is performed
  * ieframe.dll!CConnectionPoint::UnadviseAll(void) tries to access a pointer in what looks like freed memory and causes the exception

Stack trace:
<pre>
 	ieframe.dll!CConnectionPoint::UnadviseAll(void)	Unknown
 	ieframe.dll!CShellOcx::~CShellOcx()	Unknown
 	ieframe.dll!CWebBrowserOC::`vector deleting destructor'(unsigned int)	Unknown
 	ieframe.dll!CAggregatedUnknown<3>::CUnkInner::Release(void)	Unknown
 	ieframe.dll!CWebBrowserOC::Release()	Unknown
 	ieframe.dll!CIEFrameAuto::CMimeTypes::Release(void)	Unknown
 	mshtml.dll!CPlugins::~CPlugins(void)	Unknown
 	mshtml.dll!CPlugins::`vector deleting destructor'(unsigned int)	Unknown
 	mshtml.dll!CBase::SubRelease()	Unknown
 	mshtml.dll!CGarbageTracker::DeferredRelease(unsigned long)	Unknown
 	mshtml.dll!GlobalWndOnMethodCall()	Unknown
 	mshtml.dll!GlobalWndProc()	Unknown
 	user32.dll!__InternalCallWinProc@20()	Unknown
 	user32.dll!UserCallWinProcCheckWow()	Unknown
 	user32.dll!DispatchMessageWorker()	Unknown
 	user32.dll!_DispatchMessageW@4()	Unknown
>	example-html.exe!wWinMain(HINSTANCE__ * hInstance=0x00310000, HINSTANCE__ * hPrevInstance=0x00000000, wchar_t * lpCmdLine=0x00a11994, int nCmdShow=1) Line 320	C
 	example-html.exe!invoke_main() Line 118	C++
 	example-html.exe!__scrt_common_main_seh() Line 253	C++
 	example-html.exe!__scrt_common_main() Line 296	C++
 	example-html.exe!wWinMainCRTStartup() Line 17	C++
 	kernel32.dll!@BaseThreadInitThunk@12()	Unknown
 	ntdll.dll!77ee5de3()	Unknown
 	[Frames below may be incorrect and/or missing, no symbols loaded for ntdll.dll]	
 	ntdll.dll!77ee5dae()	Unknown
</pre>

<pre>
CConnectionPoint::UnadviseAll:
0F9B0F6F 8B FF                mov         edi,edi  
0F9B0F71 55                   push        ebp  
0F9B0F72 8B EC                mov         ebp,esp  
0F9B0F74 51                   push        ecx  
0F9B0F75 51                   push        ecx  
0F9B0F76 53                   push        ebx  
0F9B0F77 8B D9                mov         ebx,ecx  
0F9B0F79 56                   push        esi  
0F9B0F7A 57                   push        edi  
0F9B0F7B 83 7B 04 00          cmp         dword ptr [ebx+4],0  
0F9B0F7F 75 09                jne         CConnectionPoint::UnadviseAll+1Bh (0F9B0F8Ah)  
0F9B0F81 5F                   pop         edi  
0F9B0F82 5E                   pop         esi  
0F9B0F83 33 C0                xor         eax,eax  
0F9B0F85 5B                   pop         ebx  
0F9B0F86 8B E5                mov         esp,ebp  
0F9B0F88 5D                   pop         ebp  
0F9B0F89 C3                   ret  
0F9B0F8A 83 65 FC 00          and         dword ptr [ebp-4],0  
0F9B0F8E 83 7B 0C 00          cmp         dword ptr [ebx+0Ch],0  
0F9B0F92 76 36                jbe         CConnectionPoint::UnadviseAll+5Bh (0F9B0FCAh)  
0F9B0F94 33 C9                xor         ecx,ecx  
0F9B0F96 89 4D F8             mov         dword ptr [ebp-8],ecx  
0F9B0F99 8B 43 04             mov         eax,dword ptr [ebx+4]  
0F9B0F9C 03 C1                add         eax,ecx  
0F9B0F9E 8B 08                mov         ecx,dword ptr [eax]  
0F9B0FA0 85 C9                test        ecx,ecx  
0F9B0FA2 75 2C                jne         CConnectionPoint::UnadviseAll+61h (0F9B0FD0h)  
0F9B0FA4 8B 7B 04             mov         edi,dword ptr [ebx+4]  
0F9B0FA7 BE 58 5F 86 0F       mov         esi,offset _GUID_NULL (0F865F58h)  
0F9B0FAC 8B 4D F8             mov         ecx,dword ptr [ebp-8]  
0F9B0FAF 83 C7 08             add         edi,8  
0F9B0FB2 03 F9                add         edi,ecx  
0F9B0FB4 83 C1 18             add         ecx,18h  
0F9B0FB7 89 4D F8             mov         dword ptr [ebp-8],ecx  
0F9B0FBA A5                   movs        dword ptr es:[edi],dword ptr [esi]  
0F9B0FBB A5                   movs        dword ptr es:[edi],dword ptr [esi]  
0F9B0FBC A5                   movs        dword ptr es:[edi],dword ptr [esi]  
0F9B0FBD A5                   movs        dword ptr es:[edi],dword ptr [esi]  
0F9B0FBE 8B 75 FC             mov         esi,dword ptr [ebp-4]  
0F9B0FC1 46                   inc         esi  
0F9B0FC2 89 75 FC             mov         dword ptr [ebp-4],esi  
0F9B0FC5 3B 73 0C             cmp         esi,dword ptr [ebx+0Ch]  
0F9B0FC8 72 CF                jb          CConnectionPoint::UnadviseAll+2Ah (0F9B0F99h)  
0F9B0FCA 83 63 08 00          and         dword ptr [ebx+8],0  
0F9B0FCE EB B1                jmp         CConnectionPoint::UnadviseAll+12h (0F9B0F81h)  
0F9B0FD0 83 20 00             and         dword ptr [eax],0  
0F9B0FD3 8B 01                mov         eax,dword ptr [ecx]  
> 0F9B0FD5 8B 78 08             mov         edi,dword ptr [eax+8]  
0F9B0FD8 81 FF 00 24 9B 0F    cmp         edi,offset CTabEventProxy::Release (0F9B2400h)  
0F9B0FDE 74 1A                je          CConnectionPoint::UnadviseAll+8Bh (0F9B0FFAh)  
0F9B0FE0 8B F4                mov         esi,esp 
</pre> 

		EAX	DDDDDDDD	
		ECX	04FC6EF8	
		EDI	0F8611E0	

The biggest clue for me is mshtml.dll!CGarbageTracker::DeferredRelease(unsigned long) and the dummy_AddRef/Release implementation in the html object. I believe some references to interfaces are being kept by the internals of IE after the window is closed and the html object is freed.

I tried to implement proper AddRef/Release methods for the html object and this did seem to stop the crash (also attached), however, sometimes the html object doesn't seem to get enough releases and is not freed.
[example.zip](https://github.com/mity/mctrl/files/541665/example.zip)
[attempted-fix.zip](https://github.com/mity/mctrl/files/541666/attempted-fix.zip)

**Comment by mity** [comment]
Thanks for the report. I'll try to take a look at it tomorrow.

**Comment by dlaundon** [comment]
That's great, thank you.

Happy to help debug/test the issue if you have trouble replicating it.

**Comment by mity** [comment]
Could you also provide the modified `doc.html`? The attached `example.zip` does not contain it and I failed to reproduce without it. Given how many other components/COM object may MSIE create/activate internally for some flash or HTML5 video or whatever it uses, it may have an impact whether it can be reproduced.

**Comment by mity** [comment]
I am not able to reproduce the issue even when I tried to embed a youtube video on the `doc.html` and going "back" in the modified example while playing it. I was even unable to reproduce the leak you can see.

I added the proper ref. counting and also way more carefully crafted code to destruct the embedded browser, as I have found this: http://stackoverflow.com/a/14652605/917880

Given the code is now much cleaner, I committed even without reproducing the bug. But please test and report back. With some luck the leak you can see could be fixed with the new `html_destroy()`.

**Comment by dlaundon** [comment]
Sorry, I included the wrong file. I've attached updated example for reference.
[example.zip](https://github.com/mity/mctrl/files/543857/example.zip)

I am not surprised that you were unable to reproduce it. It has been months before I was able to do so myself, and it seems to depend on subtle timing and allocated memory layout, and which way the wind is blowing (as we say here...)

I will try the new version and report back, thank you very much for your help.
